File: WdtcCore/src/main/java/org/wdt/wdtc/core/launch/GameJvmCommand.java
package org.wdt.wdtc.core.launch;

import com.google.gson.JsonElement;
import org.wdt.utils.io.FileUtils;
import org.wdt.wdtc.core.game.GameVersionJsonObject;
import org.wdt.wdtc.core.game.Launcher;
import org.wdt.wdtc.core.game.config.DefaultGameConfig;
import org.wdt.wdtc.core.manger.VMManger;

import java.io.IOException;
import java.util.Map;

public class GameJvmCommand extends AbstractGameCommand {
  private final Launcher launcher;
  private final String LibraryList;

  public GameJvmCommand(Launcher launcher) {
    this.launcher = launcher;
    this.LibraryList = new GameClassPath(launcher).getCommand().toString();
  }

  public StringBuilder getCommand() throws IOException {
    DefaultGameConfig.Config gameConfig = launcher.getGameConfig().getConfig();
    GameVersionJsonObject VersionJsonObject = launcher.getGameVersionJsonObject();
    command.append("@echo off\n").append("cd ").append(launcher.getVersionPath()).append("\n");
    nonBreakingSpace("\"" + gameConfig.getJavaPath() + "\"");
    nonBreakingSpace("-Dlog4j.configurationFile=", launcher.getVersionLog4j2());
    nonBreakingSpace("-Xmx" + gameConfig.getMemory(), "M");
    nonBreakingSpace("-Dminecraft.client.jar=", launcher.getVersionJar());
    nonBreakingSpace("-XX:+UnlockExperimentalVMOptions -XX:+UseG1GC -XX:G1NewSizePercent=20 -XX:G1ReservePercent=20 -XX:MaxGCPauseMillis=50 -XX:G1HeapRegionSize=32m");
    nonBreakingSpace("-XX:HeapDumpPath=MojangTricksIntelDriversForPerformance_javaw.exe_minecraft.exe.heapdump");
    FileUtils.createDirectories(launcher.getVersionNativesPath());
    for (JsonElement Element : VersionJsonObject.getArguments().getJvmList()) {
      if (!Element.isJsonObject()) {
        nonBreakingSpace(ReplaceData(Element.getAsString()));
      }
    }
    return command;
  }

  private Map<String, String> getDataMap() {
    return Map.of("${natives_directory}", FileUtils.getCanonicalPath(launcher.getVersionNativesPath()), "${launcher_name}", "Wdtc",
        "${launcher_version}", VMManger.getLauncherVersion(), "${library_directory}", FileUtils.getCanonicalPath(launcher.getGameLibraryDirectory()),
        "${classpath_separator}", ";", "${version_name}", launcher.getVersionNumber(),
        "${classpath}", LibraryList);
  }

  private String ReplaceData(String str) {
    Map<String, String> ReplaceMap = getDataMap();
    for (String s : ReplaceMap.keySet()) {
      str = str.replace(s, ReplaceMap.get(s));
    }
    return str;
  }
}

File: WdtcCore/src/main/kotlin/org/wdt/wdtc/core/launch/GameJvmCommand.kt
package org.wdt.wdtc.core.launch

import org.wdt.wdtc.core.game.Version
import org.wdt.wdtc.core.manger.gameConfig
import org.wdt.wdtc.core.manger.launcherVersion
import org.wdt.wdtc.core.utils.noNull
import java.io.IOException

class GameJvmCommand(private val version: Version) : AbstractGameCommand() {
  private val libraryList: String = GameRuntimeCommnad(version).getCommand().toString()

  @Throws(IOException::class)
  override fun getCommand(): StringBuilder {
    val gameConfig = version.gameConfig.config
	  commandBuilder.append("@echo off").appendLine().append("cd ").append(version.versionDirectory).appendLine()
	  nonBreakingSpace("\"${gameConfig.javaPath}\"")
    nonBreakingSpace("-Dlog4j.configurationFile=", version.versionLog4j2)
    nonBreakingSpace("-Xmx${gameConfig.memory}", "M")
    nonBreakingSpace("-Dminecraft.client.jar=", version.versionJar)
    nonBreakingSpace("-XX:+UnlockExperimentalVMOptions -XX:+UseG1GC -XX:G1NewSizePercent=20 -XX:G1ReservePercent=20 -XX:MaxGCPauseMillis=50 -XX:G1HeapRegionSize=32m")
    nonBreakingSpace("-XX:HeapDumpPath=MojangTricksIntelDriversForPerformance_javaw.exe_minecraft.exe.heapdump")
    version.gameVersionJsonObject.arguments.jvmList.noNull().forEach {
      if (it.isJsonPrimitive) {
        nonBreakingSpace(it.asString.replaceData(dataMap))
      }
    }
    return commandBuilder
  }

  override val dataMap
    get() = mapOf(
      "natives_directory" to version.versionNativesPath.canonicalPath,
      "launcher_name" to "Wdtc",
      "launcher_version" to launcherVersion,
      "library_directory" to version.gameLibraryDirectory.canonicalPath,
      "classpath_separator" to ";",
      "version_name" to version.versionNumber,
      "classpath" to libraryList
    )

}
